design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+support `ROWTIME` in pull queries because the state stores from which pull queries are 
+served includes the timestamp for the latest table row (for each key). In contrast, 
+message partition and offset are not saved in the state stores. As a result, supporting 
+`ROWPARTITION` and `ROWOFFSET` in pull queries would require adding this information
+to state stores. (I think there might be a way to get the partition by other means,
+from the replica queried in order to serve the pull query, but offset information
+definitely wouldn't be available.)
+
+If a user really wants to make partition and/or offset information available for pull
+queries, they can choose to explicitly include the source partition and offset as
+part of the query that materializes their table and query for partition/offset as
+value columns as part of their pull query.
+
+### Other semantics
+
+Similar to `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will not 
+be included in the list of columns returned as part of a source description.
+
+`ROWPARTITION` and `ROWOFFSET` will also not be added to the output of `PRINT TOPIC` commands.
+
+Unlike `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will not
+be allowed as part of `INSERT VALUES` statements. It does not make sense for users
+to specify the offset, and while it's technically possible for users to specify a
+custom partition, this does not seem useful as violating expected partitioning has
+implications for downstream queries.
+
+## Design
+
+`ROWPARTITION` and `ROWOFFSET` will be added as new pseudocolumns, similar to how
+`ROWTIME` is already implemented today. New logic will be added to disallow the use of
+`ROWPARTITION` and `ROWOFFSET` in pull queries, while `ROWTIME` will still be allowed.
+
+However, directly adding `ROWPARTITION` and `ROWOFFSET` as new pseudocolumns would break 
+compatibility for certain existing queries since, in some situations, 
+[ksqlDB writes pseudocolumns into repartition and changelog topics](https://github.com/confluentinc/ksql/issues/7489). 
+This is problematic because adding new pseudocolumns would change the schema of these 
+intermediate topics (and state stores) and break running queries as a result.
+
+In light of the above, we have two approaches for implementation:
+1. Update ksqlDB to never write pseudocolumns into repartition and changelog topics 
+   (unless they are needed as part of the query downstream), i.e., address 
+   https://github.com/confluentinc/ksql/issues/7489. Then we can freely add new 
+   pseudocolumns without breaking compatibility for existing queries.
+2. Defer on updating ksqlDB to not write pseudocolumns into intermediate topics and 
+   instead modify the stream and table source execution steps to include something 
+   to indicate the set of pseudocolumns that should be included with the source. 
+   Specifically, we can add a "pseudocolumn version number" field into the execution 
```

Comment:
instead of a version number, why not just list out all the psuedocolumns to include? magic versions numbers can be tough to reason about

Reply:
I considered this but felt that a version number would allow for more evolvability in the future. For example, if we wanted to change the implementation of a pseudocolumn, that would be possible with a version number but not a list of column names. (A bit contrived but maybe we'll introduce `ROWHEADERS` in the future with type `ARRAY<STRUCT<key STRING, header BYTES>>` and later want to change it to a different type.)

I don't feel strongly though. Happy to keep it readable if introducing a version number feels like over-engineering.